### PR TITLE
EZEE-2303: Impl. RichText block

### PR DESCRIPTION
**Pull request by adamwojs**
| Question      | Answer
| ------------- | ---
| Tickets       | [EZEE-2303](https://jira.ez.no/browse/EZEE-2303)
| Depends on | ezsystems/ezplatform-richtext#13 |
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Tests pass?   | yes
| Doc needed?   | no
| License       | [GPL-2.0](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/LICENSE)

Refactored the RichText field type to be able to reuse it in the RichText block. 

#### Checklist:
- [x] Coding standards (`$ composer fix-cs`)
- [x] Ready for Code Review

### CODE REVIEW

File: src/bundle/Resources/views/form_fields.html.twig [hunk 1]
```diff
@@ -153,3 +153,18 @@
         {{ block('form_widget') }}
     </div>
 {%- endblock -%}
+
+{%- block richtext_widget -%}
+    {% set attr = attr|merge({ 'hidden': true }) %}
```
Comment (tischsoic) [inline src/bundle/Resources/views/form_fields.html.twig]: How does this work, do we use `attr`?

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
+(function(global) {
+    const eZ = (global.eZ = global.eZ || {});
+
+    const BaseRichtText = class BaseRichtText {
```
Comment (tischsoic) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: BaseRich\*t\*Text :)

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                    this.xhtmlNamespace,
+                    this.ezNamespace
+                );
+            });
+
+            return alloyEditor;
+        }
+    };
+
+    eZ.BaseRichtText = BaseRichtText;
```
Comment (tischsoic) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: Maybe we should use `eZ.addConfig` here?

File: src/bundle/Resources/public/js/alloyeditor/src/buttons/ez-btn-linkedit.js [hunk 4]
```diff
@@ -78,24 +78,24 @@ export default class EzBtnLinkEdit extends Component {
...
-                        },
-                        config
-                    )
-                ),
-                udwContainer
+            const alloyEditorCallbacks = eZ.ezAlloyEditor.callbacks;
+            const mergedConfig = Object.assign(
+                {
+                    onConfirm: udwOnConfirm,
+                    onCancel: () => ReactDOM.unmountComponentAtNode(udwContainer),
```
Comment (tischsoic) [inline src/bundle/Resources/public/js/alloyeditor/src/buttons/ez-btn-linkedit.js]: I wonder, should we pass this in config even if we are using `alloyEditorCallbacks.openUdw`? Probably there is no need to unmount component at `udwContainer` in this case. What do you think?

File: src/bundle/Resources/views/form_fields.html.twig (hunk 1 above)
Comment (adamwojs) [reply src/bundle/Resources/views/form_fields.html.twig]: Yes, we do. Is used by the `widget_attributes` block called in https://github.com/ezsystems/ezplatform-admin-ui/pull/714/files#diff-c5ade6303c80a553058d3c44229e563bR165

File: src/bundle/Resources/views/form_fields.html.twig [hunk 5]
```diff
@@ -153,3 +153,18 @@
...
+    {# FIXME: hardcoded language code #}
```
[the remark this answers is not in the document]
Comment (adamwojs) [reply src/bundle/Resources/views/form_fields.html.twig]: No, it's not supposed to stay. I will fix it.

File: src/bundle/Resources/views/form_fields.html.twig (hunk 1 above)
Comment (tischsoic) [reply src/bundle/Resources/views/form_fields.html.twig]: Great, thanks for explanation. :)

File: src/bundle/Resources/public/js/alloyeditor/src/buttons/ez-btn-linkedit.js (hunk 4 above)
Comment (dew326) [reply src/bundle/Resources/public/js/alloyeditor/src/buttons/ez-btn-linkedit.js]: This is for a more generic solution. Someone can set own callback to openUDW and may rely on this method.

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js (hunk 3 above)
Comment (dew326) [reply src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: I want to be in line with the rest of the AE files.

**Comment by sunpietro** [comment]
There are conflicts in this branch @adamwojs

**Comment by adamwojs** [comment]
PR updated according to the @tischsoic suggestions (I also squashed the commits)

**Comment by adamwojs** [comment]
@webhdx My point in placing mentioned classes in the `ezplatform-admin-ui` repository was to allow the eZ Platform Developer to use RichText as any other front-end component in their custom administration controllers. 

Place where this feature might be useful is an e-mail templates editor (screenshot from demo prepared for test purposes) zrzut ekranu 2018-11-22 o 12 54 51
or CRUD of custom doctrine entity.  

But in general this feature is something out of scope, so I won't push for it (moving it to the `ezplatform-page-builder` might event simplify things). WDYT ?

File: src/bundle/Resources/views/form_fields.html.twig [hunk 6]
```diff
@@ -153,3 +153,14 @@
...
     </div>
 {%- endblock -%}
+
+{%- block richtext_widget -%}
+    {% set attr = attr|merge({ 'hidden': true }) %}
+
+    <div class="ez-data-source">
+        <textarea {{ block('widget_attributes') }}>{{ value }}</textarea>
+        <div class="ez-data-source__richtext" id="{{ form.vars.id }}__editable"></div>
+        <div class="hidden" data-udw-config-name="richtext_embed" data-udw-config="{{ ez_udw_config('richtext_embed', udw_context) }}"></div>
```
Comment (mikadamczyk) [inline src/bundle/Resources/views/form_fields.html.twig]: nitpick: `richtext_embed` could be extracted to variable

**Pull request merged**